tests/neural_networks/test_qcnn.py

```diff
@@ -140,3 +156,39 @@ def test_sequence(self, num_qubits, operations):
                 or mock_quantum_pooling_layer3.call_count > 0
                 or mock_fully_connected_layer.call_count > 0
             )
+
+    # Tests added for BaseLayer inheritance validation
+    @pytest.mark.parametrize(
+        "num_qubits, operations",
+        [
+            (2, [(print, {})]),
+            (3, [(sum, {})]),
+            (4, [(lambda x: x, {})]),
+        ],
+    )
+    def test_non_class_operations(self, num_qubits, operations):
+        """Tests that non-class callables are rejected."""
+        with raises(TypeError, match="must be a class"):
+            _ = QCNN(num_qubits).sequence(operations)
+
+    @pytest.mark.parametrize(
+        "num_qubits, operations",
+        [
+            (2, [(dict, {})]),
+            (3, [(str, {})]),
+            (4, [(object, {})]),
+        ],
+    )
+    def test_non_baselayer_operations(self, num_qubits, operations):
+        """Tests that classes not inheriting from BaseLayer are rejected."""
+        with raises(TypeError, match="must inherit from BaseLayer"):
```

Comment:
The match argument must contain the exact error message that we are looking for. Use regex here as well.

```suggestion
        with raises(TypeError, match=r"Operation at index \d+\ must inherit from BaseLayer, got \*?"):
```